### PR TITLE
feat: Enhance scap for multi-display screen capturing (Mac OS)

### PR DESCRIPTION
**Pull request by vasusharma7**
This is pre-requisite for https://github.com/helmerapp/micro/issues/59 and enhances following things in scap:

 * add get_display method to access a display with particular id
 * add display id screen capture options
 * expose get_all_displays method from scap for Mac OS

### CODE REVIEW

**Comment by vasusharma7** [comment]
An observation I had on my device:

Often times, the control flow gets stuck at `SCShareableContent::current()` and `CGDisplay::bounds(..)` function calls when I run scap directly (main.rs testbed for scap). I haven't been able to figure out the reasons behind this.

But this doesn't happen when scap exposed methods ( `get_all_displays()`, `get_targets()` etc.) are invoked from within micro ( which internally call `SCShareableContent::current()`, `CGDisplay::bounds(..)` )

File: src/lib.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ mod device;
 pub mod frame;
 
 // Helper Methods
+pub use device::display::get_all_displays;
 pub use device::display::get_targets;
 pub use device::display::has_permission;
```
Comment (clearlysid) [inline src/lib.rs]: imo we can skip exposing `get_all_displays` here. We should keep our API footprint small and let people call `get_targets` and then filter out what they need from there itself.

**Comment by clearlysid** [comment]
Great PR, thanks!

I think one change I'd like is we to make a "Target" enum struct that wraps "Displays" and "Windows".

**Comment by vasusharma7** [comment]
Hi @clearlysid 

Not sure if I fully understood your intent here ( being new to rust & this space in general ), but I have tried to do the changes in recent commit.

Overview:

* Targets is now an enum struct wrapping Displays and Windows
* Modified get_targets to return concatenation of get_windows and get_displays
* Made device module public to consume Targets enum from micro and other library users
* Re-used 'Point' and 'Size' structs from capturer module to convey size and physical position of displays

 PTAL & provide feedback :)

File: src/lib.rs (hunk 1 above)
Comment (vasusharma7) [reply src/lib.rs]: Done

**Comment by Pranav2612000** [comment]
> Great PR, thanks!
> 
> I think one change I'd like is we to make a "Target" enum struct that wraps "Displays" and "Windows".

And Applications ( https://developer.apple.com/documentation/screencapturekit/scrunningapplication ) as well? @clearlysid 

This'll be different for each OS, so we'll need to create our own common structs.

**Comment by clearlysid** [comment]
@Pranav2612000 check PR #78 -- I made some progress on this.

> And Applications ( https://developer.apple.com/documentation/screencapturekit/scrunningapplication ) as well?

Applications isn't a requirement for Helmer atm so I'm not personally prioritizing it, but can consider adding later if there is some demand for it.

**Comment by clearlysid** [comment]
@vasusharma7 check PR #78, I have added the Target struct and updated `scap::get_targets()` to return a list of all capture-able entites. That way users can filter out things from there to pass as `targets`.

Now: instead of passing an explicit `display_id` in the options, we can just iterate over the given `targets` to decide what display to capture. Wdyt?

**Comment by vasusharma7** [comment]
Didn't look at 100% of the PR but I see that you have exposed raw_handle for CGDisplay among other changes. Overall LGTM!

That PR seems to cover all the things done here in this PR.

**Comment by clearlysid** [comment]
> That PR seems to cover all the things done here in this PR.

Not all 😬 -- I've exposed all monitors in targets, but currently the capturer doesn't use the targets and only captures the primary display. Do you think you can carry over your capturer functionality to that PR (or merge that one here and build on top?)

That way, you can do this for macOS and I will finish it for Windows. What say?

**Comment by clearlysid** [comment]
This TODO line is what I'm asking for help with, basically:
https://github.com/helmerapp/scap/pull/78/files#diff-d3867342fe04a20c00d4a70e7d3fcc4e349ec0dc32274afc0cd22b414dac087fR99

**Comment by vasusharma7** [comment]
Sorry, didnt have a thorough look 😅
Yes, I will continue on the capturer in that PR only

**Comment by vasusharma7** [comment]
Didn't get time to work on this today , will complete this soon .....

**Comment by clearlysid** [comment]
Have solved this as part of #87

**Comment by Pranav2612000** [comment]
@all-contributors please add @vasusharma7  for coding

**Pull request closed without merging**